### PR TITLE
feat(eslint-plugin): [unified-signatures] add `ignoreDifferentlyNamedParameters` option

### LINKED ISSUE #2386: [unified-signatures] Don't trigger on mergeable declaration with different parameter name
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/unified-signatures": "error"
  }
}
```

```TS
declare const findUp: {
	(name: string | readonly string[], options?: findUp.Options): Promise<string | undefined>;

	(matcher: (directory: string) => (findUp.Match | Promise<findUp.Match>), options?: findUp.Options): Promise<string | undefined>;
};
```

**Expected Result**

I expected it not to be a violation as a different parameter name is a good indication that the overloads are not compatible.

**Actual Result**

```
  ✖   57:4  These overloads can be combined into one signature taking string | string[] | (directory: string) => findUp.Match.                            @typescript-eslint/unified-signatures
  ✖  134:3  These overloads can be combined into one signature taking string | string[] | (directory: string) => (findUp.Match | Promise<findUp.Match>).  @typescript-eslint/unified-signatures
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.9.0` |
| `@typescript-eslint/parser`        | `3.9.0` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.6.0` |
| `node`                             | `10.22.0` |
| `npm`                              | `6.14.7` |

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

-   [x] Addresses an existing issue: fixes #2386
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

When two signatures have the same number of parameters, then two parameters having a different name means they won't be considered unifiable.

Also deduplicates a bit of code in `checkOverloads`.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 1]
```diff
@@ -140,35 +165,19 @@ export default util.createRule({
...
       const isTypeParameter = getIsTypeParameter(typeParameters);
       for (const overloads of signatures) {
-        if (overloads.length === 2) {
-          const signature0 =
-            (overloads[0] as MethodDefinition).value || overloads[0];
-          const signature1 =
-            (overloads[1] as MethodDefinition).value || overloads[1];
+        forEachPair(overloads, (a, b) => {
+          const signature0 = (a as MethodDefinition).value || a;
+          const signature1 = (b as MethodDefinition).value || b;
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/unified-signatures.ts:170]:
```suggestion
          const signature0 = (a as MethodDefinition).value ?? a;
          const signature1 = (b as MethodDefinition).value ?? b;
```

**Review by bradzacher — CHANGES_REQUESTED** [review]
this is looking good to me so far!
I didn't see any test cases for it - could you please add a negative test with the option `false`?

**Comment by JoshuaKGoldberg** [comment]
Heh, turns out the rule already wasn't testing one of the positive test cases. Classic.

File: packages/eslint-plugin/src/rules/unified-signatures.ts [hunk 2]
```diff
@@ -591,6 +615,16 @@ function getOverloadInfo(node: OverloadNode): string {
   }
 }
 
+function getStaticParameterName(param: TSESTree.Node): string | undefined {
+  switch (param.type) {
+    case AST_NODE_TYPES.Identifier:
+      return param.name;
+    case AST_NODE_TYPES.RestElement:
+      return getStaticParameterName(param.argument);
+    default:
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/unified-signatures.ts:624]: quick note that these is valid:
```ts
function foo({a}: {a: string}): void;
function foo([a]: [a: string]): void;
```

I think it's fine for us to ignore these cases as they're rare and pretty dumb.

**Review by bradzacher — APPROVED** [review]
LGTM! Thanks for adding this!

**Pull request merged**